### PR TITLE
streamclient: add random stream client

### PR DESCRIPTION
**Pull request by pbardea**
See individual commits, but this PR does some cleanup while introducing a
stream client implementation that randomly generates rows.

I broke down each change into its own commit for reviews, but let me know
if splitting this into separate PRs would be helpful.

### CODE REVIEW

**Review by adityamaru — APPROVED** [review]
> Reviewed 18 of 18 files at r1, 7 of 7 files at r2, 5 of 5 files at r3, 10 of 10 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59139)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/streamingccl/streamclient/client.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQAp4-752ODX1_F_bV:-MRQAp4-752ODX1_F_bW:bg1l9f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d085b0915745a8d6612de2b6db2d06d49bfb9f9c/pkg/ccl/streamingccl/streamclient/client.go#L30)):*
> ```
> the partition and close the event channel
> ```
nit: last part of the sentence is repeated.
___
*[pkg/ccl/streamingccl/streamclient/client_test.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQ_Faw0gIkl_FcTgsZ:-MRQ_FaxDd26--igaOtq:b-ez0udl) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ed0fd1bf61ddeb80a7d4d964274e15502cd2a/pkg/ccl/streamingccl/streamclient/client_test.go#L89)):*
> ```Go
> 				panic(fmt.Sprintf("unexpected event type %v", event.Type()))
> 			}
> 			numReceivedEvents++
> ```

can we delete?
___
*[pkg/ccl/streamingccl/streamclient/random_stream_client.go, line 186 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQLNjQ4pQZmu8r4i94:-MRQLNjQ4pQZmu8r4i95:badamgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37f152777af08556da6f67d689929c4b7b11da8/pkg/ccl/streamingccl/streamclient/random_stream_client.go#L186)):*
> ```Go
> 			}
> 
> 			// TODO: Consider keeping an in-memory copy so that tests can verify
> ```

+1 this was on my mind too.
___
*[pkg/ccl/streamingccl/streamclient/random_stream_client.go, line 199 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQM2J54pBiF-CUxqg4:-MRQM2J54pBiF-CUxqg5:beuet5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37f152777af08556da6f67d689929c4b7b11da8/pkg/ccl/streamingccl/streamclient/random_stream_client.go#L199)):*
> ```Go
> }
> 
> func (m *randomStreamClient) makeRandomKey(r *rand.Rand, minTs time.Time) roachpb.KeyValue {
> ```

love this new version!
___
*[pkg/kv/bulk/sst_batcher.go, line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQZNtnEHaO09R8Ix1x:-MRQZNtoDt4W912rNAoS:b-jg6ewz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37f152777af08556da6f67d689929c4b7b11da8/pkg/kv/bulk/sst_batcher.go#L146)):*
> ```Go
> // Keys must be added in order.
> func (b *SSTBatcher) AddMVCCKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
> 	if len(b.batchEndKey) > 0 && bytes.Equal(b.batchEndKey, key.Key) && b.disallowShadowing {
> ```

nice catch

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go [hunk 1]
```diff
@@ -241,28 +210,25 @@ func (sip *streamIngestionProcessor) flush() error {
...
-					if !ok {
-						return
-					}
-					merged <- partitionEvent{
-						Event:     event,
-						partition: partition,
-					}
+				case merged <- pe:
 				case <-ctx.Done():
-					return
```
Comment (pbardea) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:265]: Note for me: add this `return` back.

File: pkg/ccl/streamingccl/streamclient/stream_client.go [hunk 2]
```diff
@@ -8,29 +8,35 @@
...
-func (m *client) ConsumePartition(
-	address PartitionAddress, startTime time.Time,
-) (chan Event, error) {
-	eventCh := make(chan Event)
+func (m *mockClient) ConsumePartition(
+	ctx context.Context, _ streamingccl.PartitionAddress, _ time.Time,
+) (chan streamingccl.Event, error) {
+	eventCh := make(chan streamingccl.Event)
+	go func() {
+		select {
```
Comment (pbardea) [inline pkg/ccl/streamingccl/streamclient/stream_client.go:36]: This doesn't need to be a select.

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59139)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/streamclient/client.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQAp4-752ODX1_F_bV:-MRQmUBp47u6-aPafDoW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d085b0915745a8d6612de2b6db2d06d49bfb9f9c/pkg/ccl/streamingccl/streamclient/client.go#L30)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

> ```
> the partition and close the event channel
> ```
nit: last part of the sentence is repeated.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamclient/client_test.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQ_Faw0gIkl_FcTgsZ:-MRQmSgY-gvbRg_L1f2n:b-4nowv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ed0fd1bf61ddeb80a7d4d964274e15502cd2a/pkg/ccl/streamingccl/streamclient/client_test.go#L89)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

can we delete?
</blockquote></details>

Nice catch. Done.
___
*[pkg/ccl/streamingccl/streamclient/stream_client.go, line 36 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQ2_9CJFBMzFhEe8yG-r7-36:-MRR5jNDEFAqh5TZy75m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c8c6b8099da81eed6259687f17fb2e52f38d978/pkg/ccl/streamingccl/streamclient/stream_client.go#L36)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

This doesn't need to be a select.
</blockquote></details>

Done.
___
*[pkg/kv/bulk/sst_batcher.go, line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59139#-MRQZNtnEHaO09R8Ix1x:-MRQmKlTAqshiftrv8E4:biw0u88) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37f152777af08556da6f67d689929c4b7b11da8/pkg/kv/bulk/sst_batcher.go#L146)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nice catch
</blockquote></details>

I don't think this was a bug before since we always ingested a snapshot of the data rather than the entire revision history.

**Comment by pbardea** [comment]
The tests against CI were flaking so I through together a change that allows the tests to inject "observers" into the client. Curious what you think about this approach. It allows the tests to run deterministically and we could leverage this by adding another "observer" in the test that keeps the in-memory state like we were talking about.

**Comment by adityamaru** [comment]
> The tests against CI were flaking so I through together a change that allows the tests to inject "observers" into the client. Curious what you think about this approach. It allows the tests to run deterministically and we could leverage this by adding another "observer" in the test that keeps the in-memory state like we were talking about.

Just looked at it, and it's a good idea. I guess we still rely on being able to get a hold of the client to be able to invoke `client.RegisterInterceptor`, which will be hard when we write job level tests. We'd probably still need to add a testing knob where we deposit our "interceptors" and then the ingestion processor can register those for us in the `NewStreamClient` call for `test://` URIs?

**Comment by adityamaru** [comment]
LGTM!

**Comment by pbardea** [comment]
TFTR!
bors r+

**Comment by shermanCRL** [comment]
cc @HonoreDB this might be informative, as its the client of the producer stream

**Pull request merged**